### ISSUE TITLE
Deterministic training

### ISSUE THREAD

**Issue by michaelshiyu** [NVIDIA/pix2pixHD#142]
Hi all,

I was trying to make training deterministic across runs. Here are the modifications I added at the beginning of ```train.py```:

```
torch.manual_seed(1234)
np.random.seed(1234)
torch.cuda.manual_seed(1234)
torch.backends.cudnn.deterministic = True
```

This makes the first few training epochs deterministic across runs, but as training continues, the results start to become different for different runs with the same random seed. Could anyone please help? Many thanks in advance.

**Comment by aayushARM** [comment]
@michaelshiyu Hi, this is interesting what you're trying to do. Is it actually possible to make a neural network purely deterministic? I read the code and found that there's no random noise and no random dropout either, but I'm still not sure.

**Comment by michaelshiyu** [comment]
Hi @aayushARM, it is. And for pix2pixHD, I managed to make it deterministic on CPU. But the training is still not deterministic on GPUs.

[In this discussion](https://discuss.pytorch.org/t/unable-to-obtain-deterministic-training/51111/3), it seems that there are operations in PyTorch that cannot be easily made deterministic on GPUs. I haven't looked into the details so I'm not sure if they are used in pix2pixHD or if they are the root cause for the randomness in training.

These are all purely engineering reasons, though. There's nothing preventing your net from being deterministic in theory.

Thanks, and hope this info helps!

**Comment by aayushARM** [comment]
Ok, that makes sense. Just curious, what are some scenarios where you can take advantage of determinism? I can think that it'd be useful when you want to **guarantee** reproducibility of results, but what else?

**Comment by michaelshiyu** [comment]
@aayushARM yeah reproducibility is the main thing for me. 

And maybe also debugging? Like you want to make sure a certain change you made to your codebase does what you expect. In that case, if your training results are completely deterministic, it would be easier to spot a bug. 

Thanks.
